src/main/java/com/mongodb/hibernate/jdbc/MongoStatement.java

```diff
@@ -228,21 +279,379 @@ void checkClosed() throws SQLException {
+        var writeErrors = mongoBulkWriteException.getWriteErrors();
+        var writeConcernError = mongoBulkWriteException.getWriteConcernError();
+        if (writeConcernError == null) {
+            if (!writeErrors.isEmpty()) {
+                var failedModelIndex = writeErrors.get(0).getIndex();
+                var commandIndex = writeModelsToCommandMapper.findCommandIndex(failedModelIndex);
+                return createUpdateCounts(commandIndex);
+            }
+        }
+        return EMPTY_BATCH_RESULT;
+    }
+
+    private static boolean isTimeoutException(MongoException exception) {
+        // We do not check for `MongoExecutionTimeoutException` and `MongoOperationTimeoutException` here,
+        // because it is handled via error codes.
+        return exception instanceof MongoSocketReadTimeoutException
+                || exception instanceof MongoSocketWriteTimeoutException
+                || exception instanceof MongoTimeoutException;
+    }
+
+    enum CommandDescription {
+        INSERT("insert", false, true),
+        UPDATE("update", false, true),
+        DELETE("delete", false, true),
+        AGGREGATE("aggregate", true, false);
+
+        private final String commandName;
+        private final boolean returnsResultSet;
+        private final boolean isUpdate;
+
+        CommandDescription(String commandName, boolean returnsResultSet, boolean isUpdate) {
+            this.commandName = commandName;
+            this.returnsResultSet = returnsResultSet;
+            this.isUpdate = isUpdate;
+        }
+
+        String getCommandName() {
+            return commandName;
+        }
+
+        /**
+         * Indicates whether the command is used in {@code executeUpdate(...)} or {@code executeBatch()} methods.
+         *
+         * @return true if the command is used in update operations, false if it is used in query operations.
+         */
+        boolean isUpdate() {
+            return isUpdate;
+        }
+
+        /**
+         * Indicates whether the command returns a {@link ResultSet}.
+         *
+         * @see #executeQuery(String)
+         */
+        boolean returnsResultSet() {
+            return returnsResultSet;
+        }
+
+        static CommandDescription fromString(String commandName) throws SQLFeatureNotSupportedException {
+            return switch (commandName) {
+                case "insert" -> INSERT;
+                case "update" -> UPDATE;
+                case "delete" -> DELETE;
+                case "aggregate" -> AGGREGATE;
+                default -> throw new SQLFeatureNotSupportedException("Unsupported command: %s".formatted(commandName));
+            };
+        }
+    }
+
+    private static class WriteModelConverter {
+        private static final String UNSUPPORTED_MESSAGE_STATEMENT_FIELD = "Unsupported field in %s statement: [%s]";
+        private static final String UNSUPPORTED_MESSAGE_COMMAND_FIELD = "Unsupported field in %s command: [%s]";
```

Comment:
- These are message templates, not messages.
- Shouldn't the first `%s` be also wrapped in `[`, `]`?


```suggestion
        private static final String UNSUPPORTED_MESSAGE_TEMPLATE_STATEMENT_FIELD =
                "Unsupported field in [%s] statement: [%s]";
        private static final String UNSUPPORTED_MESSAGE_TEMPLATE_COMMAND_FIELD =
                "Unsupported field in [%s] command: [%s]";
```